File: solutions/ruby/tournament/1/tournament.rb
class Tournament
  # Encabezado constante de la tabla
  HEADER = "#{'Team'.ljust(30)} | MP |  W |  D |  L |  P\n"

  def self.tally(input)
    # 1. Usamos un Hash con un bloque por defecto para crear nuevos equipos automáticamente
    teams = Hash.new { |h, k| h[k] = Team.new(k) }

    # 2. Procesamos cada línea del input
    input.strip.split("\n").each do |line|
      team1_name, team2_name, result = line.split(';')
      
      case result
      when 'win'
        teams[team1_name].add_win
        teams[team2_name].add_loss
      when 'loss'
        teams[team1_name].add_loss
        teams[team2_name].add_win
      when 'draw'
        teams[team1_name].add_draw
        teams[team2_name].add_draw
      end
    end

    # 3. Ordenamos: -points (descendente), name (alfabético)
    sorted_teams = teams.values.sort_by { |t| [-t.points, t.name] }

    # 4. Generamos la tabla final
    HEADER + sorted_teams.map(&:format_row).join("\n") + (sorted_teams.empty? ? "" : "\n")
  end

  # Clase interna para encapsular la lógica de cada equipo
  class Team
    attr_reader :name, :w, :d, :l

    def initialize(name)
      @name = name
      @w = @d = @l = 0
    end

    def mp; @w + @d + @l; end
    def points; (@w * 3) + @d; end

    def add_win;  @w += 1; end
    def add_draw; @d += 1; end
    def add_loss; @l += 1; end

    def format_row
      # %-30s -> String de 30 caracteres alineado a la izquierda
      # %3d   -> Entero de 3 espacios (para que los números se vean ordenados)
      format("%-30s | %2d | %2d | %2d | %2d | %2d", name, mp, w, d, l, points)
    end
  end
end

File: solutions/java/tournament/1/src/main/java/Tournament.java
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

class Tournament {
    // Usamos un mapa para almacenar y acceder rápidamente a cada equipo por su nombre
    private final Map<String, TeamStats> teams = new HashMap<>();

    // Plantilla de fila fija con espaciados exactos: 30 caracteres para el nombre, 
    // seguidos de las columnas numéricas alineadas a la derecha con 2 espacios de ancho.
    private static final String HEADER = "Team                           | MP |  W |  D |  L |  P\n";
    private static final String ROW_FORMAT = "%-30s | %2d | %2d | %2d | %2d | %2d\n";

    void applyResults(String resultString) {
        if (resultString == null || resultString.strip().isEmpty()) {
            return;
        }

        // Procesamos el string línea por línea
        String[] lines = resultString.split("\n");
        for (String line : lines) {
            line = line.strip();
            if (line.isEmpty()) continue;

            String[] tokens = line.split(";");
            if (tokens.length < 3) continue;

            String teamAName = tokens[0];
            String teamBName = tokens[1];
            String outcome = tokens[2];

            // Obtenemos o creamos las estadísticas para ambos equipos
            TeamStats teamA = teams.computeIfAbsent(teamAName, TeamStats::new);
            TeamStats teamB = teams.computeIfAbsent(teamBName, TeamStats::new);

            // Aplicamos las reglas del torneo según el resultado del primer equipo
            switch (outcome) {
                case "win":
                    teamA.addWin();
                    teamB.addLoss();
                    break;
                case "loss":
                    teamA.addLoss();
                    teamB.addWin();
                    break;
                case "draw":
                    teamA.addDraw();
                    teamB.addDraw();
                    break;
            }
        }
    }

    String printTable() {
        // Si no se procesó ningún partido, devolvemos únicamente la cabecera de la tabla
        if (teams.isEmpty()) {
            return HEADER;
        }

        // Ordenamos los equipos usando un comparador personalizado
        String tableBody = teams.values().stream()
            .sorted((t1, t2) -> {
                // Regla 1: Por puntos, de forma descendente (t2 comparado con t1)
                int pointsCompare = Integer.compare(t2.getPoints(), t1.getPoints());
                if (pointsCompare != 0) {
                    return pointsCompare;
                }
                // Regla 2: En caso de empate, alfabéticamente de forma ascendente (t1 comparado con t2)
                return t1.getName().compareTo(t2.getName());
            })
            // Mapeamos cada objeto de equipo a su respectivo formato de fila de texto
            .map(team -> String.format(ROW_FORMAT, 
                team.getName(), team.getMatchesPlayed(), team.getWins(), 
                team.getDraws(), team.getLosses(), team.getPoints()))
            .collect(Collectors.joining());

        return HEADER + tableBody;
    }

    // --- Clase Interna para Representar las Estadísticas de un Equipo ---
    private static class TeamStats {
        private final String name;
        private int wins = 0;
        private int draws = 0;
        private int losses = 0;

        public TeamStats(String name) {
            this.name = name;
        }

        public String getName() { return name; }
        public int getWins() { return wins; }
        public int getDraws() { return draws; }
        public int getLosses() { return losses; }

        public int getMatchesPlayed() {
            return wins + draws + losses;
        }

        public int getPoints() {
            return (wins * 3) + (draws * 1);
        }

        public void addWin() { wins++; }
        public void addDraw() { draws++; }
        public void addLoss() { losses++; }
    }
}